### PR TITLE
Nydus: refactor build workflow with packer

### PR DESCRIPTION
**Pull request by imeoer**
Introduce a new package called packer, which implements the build workflow of nydus image, as well as the export and import of build cache, the build workflow previously implemented by the nydus driver has been wrapped into
this package.

The reason for introducing packer is that nydus intends to **integrate packer as a generic component into other build toolkits such as buildkit** to more easily support building and exporting of nydus image, we have an available implementation on buildkit in this [commit](https://github.com/imeoer/buildkit/commit/1f37cb5da1f862f1771369cb93bb89c5415bf047). cc @AkihiroSuda @ktock 

image

Unlike other compression format (gzip, estargz etc.) in OCI image, nydus is divided into two types of layer, blob and bootstrap, where blob serves as the data part of each layer of the image, and bootstrap serves as the metadata of the whole image, the bootstrap is equivalent to the view of the whole image filesystem after all layers overlay. For example, for an OCI image with 3 layers, the corresponding nydus image is 4 layers (3 layers of blob + 1 layer of bootstrap), and since nydus supports chunk-level data de-duplication, it can do cross-layer or cross-image data de-duplication, so it is possible that an OCI layer does not have a corresponding nydus blob layer, because it was de-duplicated during the build workflow.

The packer also implements support for build cache, which is used to speed up the conversion or build of nydus image, i.e. when a built image layer appears in a new image, it does not need to be built again. nydus' build cache principle is to write the blob and bootstrap data of each layer to the content store at each build, and store the cache record. In the next build, will find the layers that have already been built from the cache record, skip the build of these cached layer, and reuse the cached bootstrap as parent to build a new nydus image with new layers.

image

The build cache interface (SetCache/GetCache) in packer does not implement in acceld yet, but only buildkit. Next, acceld will implement a cache manager to support build cache for each driver (nydus, estargz etc.). cc @ktock 

Nydus compression type support on buildkit: https://github.com/imeoer/buildkit/commit/1f37cb5da1f862f1771369cb93bb89c5415bf047

Signed-off-by: Yan Song <imeoer@linux.alibaba.com>

### CODE REVIEW

File: pkg/driver/nydus/packer/packer.go [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+		diffHintPaths = append(diffHintPaths, layer.diffHintPath)
+	}
+
+	var parentBootstrapPath *string
+	if diffSkip != nil {
+		// Found nydus bootstrap cache, unpack targz and use it as parent bootstrap.
+		_parentBootstrapPath := path.Join(bootstrapDir, "parent-bootstrap")
+		parentBootstrapPath = &_parentBootstrapPath
+		layer := layers[*diffSkip]
+		cachedBootstrap, _ := layer.GetCache(ctx, CompressionTypeBootstrap)
```
Comment (ktock) [inline pkg/driver/nydus/packer/packer.go:99]: error shouldn't be ignored or should have a comment at least.

File: pkg/driver/nydus/packer/packer.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+	descs := make([]Descriptor, len(layers))
+
+	// Find cache first, to skip layers that have been built.
+	buildLayers := []*BuildLayer{}
+	for idx := range layers {
+		layer := layers[idx]
+
+		// Find the layer cache.
+		cachedBootstrapDesc, _ := layer.GetCache(ctx, CompressionTypeBootstrap)
```
Comment (ktock) [inline pkg/driver/nydus/packer/packer.go:149]: error shouldn't be ignored or should have a comment at least.

File: pkg/driver/nydus/packer/mount.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
+// Copyright Project BuildKit Authors
```
Comment (ktock) [inline pkg/driver/nydus/packer/mount.go:1]: Should have a reference to the original file

Comment (imeoer) [reply pkg/driver/nydus/packer/mount.go:1]: Thanks, added a comment for this.

File: pkg/driver/nydus/packer/packer.go (hunk 1 above)
Comment (imeoer) [reply pkg/driver/nydus/packer/packer.go:102]: Added a comment.

File: pkg/driver/nydus/packer/packer.go (hunk 2 above)
Comment (imeoer) [reply pkg/driver/nydus/packer/packer.go:155]: Added a comment.

**Comment by imeoer** [comment]
Hi @AkihiroSuda @ktock 

Nydus is supporting ([pre commit](https://github.com/imeoer/buildkit/commit/c04593c9208e6cd22dd2054e83563a4790c1b5ba)) the export of image in buildkit as a compression type, currently the following points need to be noted:

1. Nydus image layers don't correspond one-to-one with OCI image layers, some OCI image layers may not have a corresponding nydus `blob` layer, and there is an additional `bootstrap` layer in nydus image.
2. Nydus image layers don't support being placed in a image manifest together with other compression types, e.g. can't use a layer combination like [tar.gz + tar.gz + nydus.blob].
3. Nydus build and runtime (https://github.com/containerd/project/issues/83) only supports linux image for now, not windows yet.

Help to see if there are any comments or suggestions? We'll submit a formal PR to buildkit next, thanks.

**Review by ktock — APPROVED** [review]

**Comment by ktock** [comment]
> Nydus image layers don't correspond one-to-one with OCI image layers, some OCI image layers may not have a corresponding nydus blob layer, and there is an additional bootstrap layer in nydus image.

Can BuildKit use the built nydus image as base image or build caches? How does it work with remote cache exporter?

**Comment by imeoer** [comment]
@ktock Thanks for the feedback.

> Can BuildKit use the built nydus image as base image or build caches?

Haven't yet, currently this [pre commit](https://github.com/imeoer/buildkit/commit/1f37cb5da1f862f1771369cb93bb89c5415bf047) only support to export nydus image, I will take a look at the implementation of remote snapshot, to support it based on nydus snapshotter next.

> How does it work with remote cache exporter?

Does the remote cache mean skipping the layer export of compressed blob already in remote registry, but not in local content store?

**Comment by ktock** [comment]
> Does the remote cache mean skipping the layer export of compressed blob already in remote registry, but not in local content store?

I meant https://github.com/moby/buildkit#export-cache

**Comment by imeoer** [comment]
> > Does the remote cache mean skipping the layer export of compressed blob already in remote registry, but not in local content store?
> 
> I meant https://github.com/moby/buildkit#export-cache

As I understand it, this cache is only applied in the build stage but not in the export stage, is that right?

**Comment by imeoer** [comment]
> I found that the inline cache is a point that nydus manifest needs to be compatible, will feedback later after some testing.

 Is still working for inline cache export.

**Comment by ktock** [comment]
https://github.com/imeoer/buildkit/commit/1f37cb5da1f862f1771369cb93bb89c5415bf047

```golang
	layer.sr.SetString(key, string(valBytes), "")
```

Nydus blob doesn't seems to be recorded as a blob in BuildKit's cache metadata? I'm a bit worry about  does it work with other parts of the cache. Could it guarantee that this works with other cache APIs like DiskUsage, Prune, etc?

**Comment by imeoer** [comment]
> [imeoer/buildkit@1f37cb5](https://github.com/imeoer/buildkit/commit/1f37cb5da1f862f1771369cb93bb89c5415bf047)
> 
> ```go
> 	layer.sr.SetString(key, string(valBytes), "")
> ```
> 
> Nydus blob doesn't seems to be recorded as a blob in BuildKit's cache metadata? I'm a bit worry about does it work with other parts of the cache. Could it guarantee that this works with other cache APIs like DiskUsage, Prune, etc?

@ktock Thanks for the reply. There is indeed this problem. Since nydus has cross-layer de-duplication, some cache records may not have nydus blob, and nydus blob cannot be mixed with other compression types in the same image.

This makes it unsuitable to record nydus blob directly to cache metadata using SetBlob instead of adding a new kv item for recording. I'll give some more thought to how to be compatible with DiskUsage.

**Comment by imeoer** [comment]
> [imeoer/buildkit@1f37cb5](https://github.com/imeoer/buildkit/commit/1f37cb5da1f862f1771369cb93bb89c5415bf047)
> 
> ```go
> 	layer.sr.SetString(key, string(valBytes), "")
> ```
> 
> Nydus blob doesn't seems to be recorded as a blob in BuildKit's cache metadata? I'm a bit worry about does it work with other parts of the cache. Could it guarantee that this works with other cache APIs like DiskUsage, Prune, etc?

@ktock [Fixed](https://github.com/imeoer/buildkit/commit/2d2dc0ef2aaa92535e28dc703a292c75295f18a1) the issue by recording nydus bootstrap to cache metadata using setBlob:

```
// Cache nydus bootstrap layer.
if err := layer.sr.setBlob(ctx, compression.Parse(packer.CompressionTypeBootstrap), desc.Bootstrap); err != nil {
	return nil, errors.Wrap(err, "set nydus bootstrap")
}

// Cache nydus blob layer.
if desc.Blob != nil {
	if err := layer.sr.addCompressionBlob(ctx, *desc.Blob, compression.Parse(packer.CompressionTypeBlob)); err != nil {
		return nil, errors.Wrap(err, "set nydus blob")
	}
}
```

If there are no more problems with this commit for now, I will submit a formal PR to buildkit next, thanks.

**Pull request merged**